### PR TITLE
Off-ff scripts as of April, 2019 (New analysis files)

### PR DESCRIPTION
**Pull request by ehrmanj**

### CODE REVIEW

File: dataextractor.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
+#Importing
```
Comment (ehrmanj) [inline dataextractor.py]: Should start with a paragraph describing what the .py does. Possibly author information, group authorship --- or not. Date last updated.

File: dataextractor.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+from rdkit import Chem, RDConfig, Geometry
```
Comment (ehrmanj) [inline dataextractor.py]: could import from most important to least important -- or most general to least general

File: dataextractor.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+import itertools
+
+
+def tanimotocombo(ref_mol, query_mol):
+    """This is the TanimotoCombo function. It takes in two OEMols. 
+    It does not matter (for our purposes) which mol is the reference mol and 
+    which one is the query mol. The TanimotoCombo distance between the two
+    mols is the same no matter which is the reference and which is the query.
+    The GetTanimotoCombo() function comes from openeye. 
+    """
```
Comment (ehrmanj) [inline dataextractor.py]: add parameters and returns to all functions

File: dataextractor.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    prep.Prep(ref_mol)
+    # Get appropriate function to calculate both shape and color
+    # By default the OEOverlapFunc contains OEGridShapeFunc for shape
+    # and OEExactColorFunc for color
+    func = oeshape.OEOverlapFunc()
+    func.SetupRef(ref_mol)
+    res = oeshape.OEOverlapResults()
+    prep.Prep(query_mol)
+    func.Overlap(query_mol, res)
+    return("%s" % res.GetTanimotoCombo())
```
Comment (ehrmanj) [inline dataextractor.py]: could be returned as float

File: dataextractor.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    This is the TFDr script. 
+    It makes use of RDKit's TFD calculation and the function rdmol_from_oemol.
+    TFDr takes in two OEMOLs. 
+    It does not matter which mol is the ref mol and which is the querymol. 
+    TFD metric is the same no matter which is the ref and which is the query.
+    First, OEmols are made RDKit compatible. Then,
+    TFD is computed and returned using RDKit's TorsionFingerprints
+    Module. Takes one input reference mol2 and one input query mol2.
+    """
+    #converts refmol to one readable by RDKit
```
Comment (ehrmanj) [inline dataextractor.py]: put a space after the hashtag

File: dataextractor.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    if (Chem.MolToSmiles(qrdmol2) != Chem.MolToSmiles(rrdmol2)):
+        tfd = -1
+    else:
+        #calculates the TFD
+        tfd = TorsionFingerprints.GetTFDBetweenMolecules(rrdmol2, qrdmol2)
+    return tfd
+
+
+
+def rdmol_from_oemol(oemol):
```
Comment (ehrmanj) [inline dataextractor.py]: should write a check to make sure the input oemol is not already RDKit-compatible

File: dataextractor.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+        tempnamedf = pd.DataFrame(np.array(list_of_molnames))
+        dname["%s" % ffdirect] = tempnamedf
+    # merges all dataframes generated above into one dataframe
+    # only keeps molecules present in all dataframes
+    twonamedf = dname[ffdirectorylist[0]].merge(dname[ffdirectorylist[1]])
+    for i in range(2, len(ffdirectorylist)):
+        threenamedf = dname[ffdirectorylist[i]].merge(twonamedf)
+    threenamedf = threenamedf.rename({0: "MolNames"},axis='columns')
+    return threenamedf
+
```
Comment (ehrmanj) [inline dataextractor.py]: Should check that at least 2 forcefields are present in the ffdirect

File: dataextractor.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+            list_of_molnames.append(molName)
+        tempnamedf = pd.DataFrame(np.array(list_of_molnames))
+        dname["%s" % ffdirect] = tempnamedf
+    # merges all dataframes generated above into one dataframe
+    # only keeps molecules present in all dataframes
+    twonamedf = dname[ffdirectorylist[0]].merge(dname[ffdirectorylist[1]])
+    for i in range(2, len(ffdirectorylist)):
+        threenamedf = dname[ffdirectorylist[i]].merge(twonamedf)
+    threenamedf = threenamedf.rename({0: "MolNames"},axis='columns')
+    return threenamedf
```
Comment (ehrmanj) [inline dataextractor.py]: all_ff_df instead of threenamedf

File: dataextractor.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+
+def all_info_df(ffdirectorylist,threenamedf):
```
Comment (ehrmanj) [inline dataextractor.py]: spaces after commas

File: dataextractor.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    for molname in threenamedf['MolNames']:
+        print(molname)
+        mol_file = '%s' % molname + '.mol2'
+        refmolin = oechem.oemolistream('%s/%s/%s' % (directory,ffdirectorylist[0],mol_file))
+        refmolhev = oechem.OEGraphMol()
+        oechem.OEReadMolecule(refmolin,refmolhev)
+        #tries to get heavyatom count -- if it can't, returns -1. 
+        try:
+            heavyvalue = oechem.OECount(refmolhev, oechem.OEIsHeavy())
+        except (Exception, ValueError):
```
Comment (ehrmanj) [inline dataextractor.py]: can probably take out the try and except 

can put check when reading in the molecule -line 227 -- if not

File: dataextractor.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+            heavyvalue = -1
+        heavyatomlist.append(heavyvalue)
+        refmolin.close()
+        #Gets TanimotoCombo and TFD values
+        for i,j in list(itertools.combinations(ffdirectorylist,2)):
+            refmolin = oechem.oemolistream('%s/%s/%s' % (directory,i, mol_file))
+            refmol = oechem.OEGraphMol()
+            oechem.OEReadMolecule(refmolin,refmol)
```
Comment (ehrmanj) [inline dataextractor.py]: Try to put heavy atoms function here so only have to open once.

File: dataextractor.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+    # and OEExactColorFunc for color
+    func = oeshape.OEOverlapFunc()
+    func.SetupRef(ref_mol)
+    res = oeshape.OEOverlapResults()
+    prep.Prep(query_mol)
+    func.Overlap(query_mol, res)
+    return("%s" % res.GetTanimotoCombo())
+
+
+def TFDr(ref_mol, query_mol):
```
Comment (ehrmanj) [inline dataextractor.py]: TFD_for_oemols(

File: dataextractor.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+            oechem.OEReadMolecule(refmolin,refmol)
+            qmolin = oechem.oemolistream('%s/%s/%s' % (directory,j, mol_file))
+            qmol = oechem.OEGraphMol()
+            oechem.OEReadMolecule(qmolin,qmol)
+            #Tries to set TFD value. If it cannot, sets as -1. 
+            try:
+                TFDvalue = TFDr(refmol,qmol)
+            except (Exception, ValueError):
+                TFDvalue = -1
+            TFDdict['%s %s' % (i, j)]['%s' % molname]=TFDvalue
```
Comment (ehrmanj) [inline dataextractor.py]: don't need to %s on the molname

File: dataextractor.py [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+            #Tries to set TanimotoCombo value. If it cannot, sets as -1. 
+            try:
+                TANIvalue = tanimotocombo(refmol,qmol)
+            except (Exception, ValueError):
+                TANIvalue = -1
+            TANIdict['%s %s' % (i, j)]['%s' % molname]=TANIvalue
+            qmolin.close()
+            refmolin.close()
+    #Loads data into dataframe
```
Comment (ehrmanj) [inline dataextractor.py]: could turn into a function

File: dataextractor.py [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+    parser.add_option('-d','--directory',
+            help = "REQUIRED: Name of the full path directory which contains the force field directories listed in --ff. If None, then the current working directory is used",
+            type = "string",
+            dest = 'directory')
+    (opt, args) = parser.parse_args()
+    if opt.ff == None:
+        parser.error("ERROR: No force fields were specified.")
+    if opt.directory == None: 
+        print("ERROR: No working directory provided.")
```
Comment (ehrmanj) [inline dataextractor.py]: parser.error

File: README.md [hunk 16]
```diff
@@ -37,7 +37,7 @@ The following is a list of the force fields being considered here:
 |OpenForceField       | 0.1.6     |
 |Schrodinger          | 2017-2  \*|
 
-\* We are working on updating this code to support the open source Desmond package provided by D.E. Shaw. This will allow for minimization with OPLS2005, whereas OPLS3 requires a license through Schrodinger. 
+\* We are working on updating this code to support the open source Desmond package provided by D.E. Shaw. This will allow for minimization with OPLS2005, whereas OPLS3 requires a license through Schrodinger.
```
Comment (davidlmobley) [inline README.md]: Are we still doing this, @ehrmanj ? If not I would remove this (and remove mention of Schrodinger in dependencies, above.)

File: dataextractor.py (hunk 3 above)
Comment (davidlmobley) [reply dataextractor.py]: You should have something in your README.md which explains what these scripts are/what they are used for/how they are used.

File: dataextractor.py [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+    # and OEExactColorFunc for color
+    func = oeshape.OEOverlapFunc()
+    func.SetupRef(ref_mol)
+    res = oeshape.OEOverlapResults()
+    prep.Prep(query_mol)
+    func.Overlap(query_mol, res)
+    return(res.GetTanimotoCombo())
+
+
+def TFD_for_oemols(ref_mol, query_mol):
```
Comment (davidlmobley) [inline dataextractor.py]: Doc string should probably explain what a TFD is and give the reference.

File: dataextractor.py (hunk 6 above)
Comment (davidlmobley) [reply dataextractor.py]: Re this comment: Create an Issue on the repo to track that you need to do this. Don't bury "to do" items in comments on GitHub as the comments become hard to find once we change the code or merge the pull request.

File: dataextractor.py (hunk 7 above)
Comment (davidlmobley) [reply dataextractor.py]: Add an Issue for this?

File: dataextractor.py (hunk 11 above)
Comment (davidlmobley) [reply dataextractor.py]: Issue

**Pull request merged**